star-aligner/src/lib.rs

```diff
@@ -109,92 +157,211 @@ impl Drop for StarAligner {
+        &mut self,
+        fqrecs: &mut [fastq::Record],
+        n_threads: usize,
+    ) -> Result<Vec<SamRecord>> {
+        // set thread pool
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(n_threads)
+            .build_global()?;
+        // split the reads into chunks by the number of threads
+        let chunk_size = (fqrecs.len() + n_threads - 1) / n_threads;
+        let fqrecs_chunks: Vec<_> = fqrecs.chunks(chunk_size).collect();
+
+        // construct an Arc mut clone to share aligner among threads
+        let aligner = Arc::new(Mutex::new(self.clone()));
+
+        // align reads in parallel
+        let records: Vec<SamRecord> = fqrecs_chunks
+            .par_iter()
+            .flat_map(|chunk| {
+                chunk
+                    .iter()
+                    .flat_map(|fqrec| {
+                        // lock the aligner
+                        let mut aligner = aligner.lock().unwrap();
+                        aligner.align_read(fqrec).unwrap()
+                    })
+                    .collect::<Vec<SamRecord>>()
+            })
+            .collect();
+
+        Ok(records)
+    }
 
-        align_read_rust(self.aligner, self.fastq1.as_slice(), &mut self.aln_buf).unwrap();
-        self.parse_sam_to_records(name)
+    // align read pairs in parallel
+    pub fn align_pair_reads(
+        &mut self,
+        fqrecs: &mut [(fastq::Record, fastq::Record)],
+        n_threads: usize,
+    ) -> Result<Vec<(SamRecord, SamRecord)>> {
+        // set thread pool
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(n_threads)
+            .build_global()?;
+        // split the reads into chunks by the number of threads
+        let chunk_size = (fqrecs.len() + n_threads - 1) / n_threads;
+        let fqrecs_chunks: Vec<_> = fqrecs.chunks(chunk_size).collect();
+
+        // construct an Arc mut clone to share aligner among threads
+        let aligner = Arc::new(Mutex::new(self.clone()));
+
+        // align reads in parallel
+        let records: Vec<(SamRecord, SamRecord)> = fqrecs_chunks
+            .par_iter()
+            .flat_map(|chunk| {
+                chunk
+                    .iter()
+                    .map(|pairfqrec| {
+                        // lock the aligner
+                        let mut aligner = aligner.lock().unwrap();
+                        aligner
+                            .align_read_pair((&pairfqrec.0, &pairfqrec.1))
+                            .unwrap()
+                    })
+                    .flat_map(|(first, second)| first.into_iter().zip(second))
+                    .collect::<Vec<(SamRecord, SamRecord)>>()
+            })
+            .collect();
+
+        Ok(records)
+    }
+
+    pub fn align_read(&mut self, fqrec: &FastqRecord) -> Result<Vec<SamRecord>> {
+        // init a aligner
+        self.aligner = unsafe { star_sys::init_aligner_from_ref(self.index.0.as_ref().index) };
+
+        let mut fq_buf = Vec::new();
```

Comment:
It may save some time if you allocate buffers outside this function, and reuse them across all reads from the same chunk.